BuildSourceImage.sh

```diff
@@ -1,80 +1,1072 @@
+    fi
+    echo -n "${base}"
+}
+
+#
+# determine, or guess, the image tag from the provided image reference
+#
+parse_img_tag() {
+    local ref="${1%@*}" # just the portion before the digest "@"
+    local tag="latest" # default tag
+
+    if [ -z "${ref}" ] ; then
+        echo -n "${tag}"
+        return 0
+    fi
+
+    local last_word="" # splitting up their reference to get the last word/chunk
+    last_word="$(echo "${ref}" | tr '/' '\n' | tail -1 )"
+    if [ "$(_count ':' "${last_word}")" -gt 0 ] ; then
+        # if there are colons in the last segment after '/', then get that tag name
+        tag="${last_word#*:}" # this parameter expansion removes the prefix pattern before the ':'
+    fi
+    echo -n "${tag}"
+}
+
+#
+# an inline prefixer for containers/image tools
+#
+ref_prefix() {
+    local ref="${1}"
+    local pfxs
+
+    # get the supported prefixes of the current version of skopeo
+    mapfile -t pfxs < <(skopeo copy --help | grep -A1 "Supported transports:" | grep -v "Supported transports" | sed 's/, /\n/g')
+
+    for pfx in "${pfxs[@]}" ; do
+        if echo "${ref}" | grep -q "^${pfx}:" ; then
+            # break if we match a known prefix
+            echo "${ref}"
+            return 0
+        fi
+    done
+    # else default
+    echo "docker://${ref}"
+}
+
+#
+# an inline namer for the source image
+# Initially this is a tagging convention (which if we try estesp/manifest-tool
+# can be directly mapped into a manifest-list/image-index).
+#
+ref_src_img_tag() {
+    local ref="${1}"
+    echo -n "$(parse_img_tag "${ref}")""${source_image_suffix}"
+}
+
+#
+# call out to registry for the image reference's digest checksum
+#
+fetch_img_digest() {
+    local ref="${1}"
+    local dgst
+    local ret
+
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    dgst=$(skopeo inspect "$(ref_prefix "${ref}")" | jq .Digest | tr -d \")
+    ret=$?
+    if [ $ret -ne 0 ] ; then
+        echo "ERROR: check the image reference: ${ref}" >&2
+        return $ret
+    fi
+
+    echo -n "${dgst}"
+}
+
+#
+# pull down the image to an OCI layout
+# arguments: image ref
+# returns: path:tag to the OCI layout
+#
+# any commands should only output to stderr, so that the caller can receive the
+# path reference to the OCI layout.
+#
+fetch_img() {
+    local ref="${1}"
+    local dst="${2}"
+    local base
+    local tag
+    local dgst
+    local from
+
+    _mkdir_p "${dst}"
+
+    base="$(parse_img_base "${ref}")"
+    tag="$(parse_img_tag "${ref}")"
+    dgst="$(parse_img_digest "${ref}")"
+    from=""
+    # skopeo currently only support _either_ tag _or_ digest, so we'll be specific.
+    if [ -n "${dgst}" ] ; then
+        from="$(ref_prefix "${base}")@${dgst}"
+    else
+        from="$(ref_prefix "${base}"):${tag}"
+    fi
+
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    ## destination name must have the image tag included (umoci expects it)
+    skopeo \
```

Comment:
Can we exit when the copy fails? Currently, we'll continue and can end up with an error as follows:
`jq: error: Could not open file /home/valentin/go/src/github.com/containers/BuildSourceImage/SrcImg/work/layouts//index.json: No such file or directory`